Fails to build with GHC 8.6

This is due to the introduction of `MonadFail`. `generateTypeVariable` contains a pattern match that can fail, but `MonadFail` isn't in the constraints:

```
src/JL/Inferer.hs:121:3: error:
    • Could not deduce (Control.Monad.Fail.MonadFail m)
        arising from a do statement
        with the failable pattern ‘v : vs’
      from the context: MonadState [TypeVariable] m
        bound by the type signature for:
                   generateTypeVariable :: forall (m :: * -> *).
                                           MonadState [TypeVariable] m =>
                                           m TypeVariable
        at src/JL/Inferer.hs:(117,1)-(119,19)
      Possible fix:
        add (Control.Monad.Fail.MonadFail m) to the context of
          the type signature for:
            generateTypeVariable :: forall (m :: * -> *).
                                    MonadState [TypeVariable] m =>
                                    m TypeVariable
    • In a stmt of a 'do' block: v : vs <- get
      In the expression:
        do v : vs <- get
           put vs
           pure v
      In an equation for ‘generateTypeVariable’:
          generateTypeVariable
            = do v : vs <- get
                 put vs
                 pure v
    |
121 |   v:vs <- get
    |   ^^^^^^^^^^^
```

--- a/src/JL/Inferer.hs
+++ b/src/JL/Inferer.hs
@@ -117,10 +117,13 @@ check ctx expr =
 generateTypeVariable
   :: MonadState ([TypeVariable]) m
   => m TypeVariable
-generateTypeVariable = do
-  v:vs <- get
-  put vs
-  pure v
+generateTypeVariable =
+  get >>= \case
+    v:vs -> do
+      put vs
+      pure v
+    _ ->
+      error "Ran out of type variables"
 
 -- | Unify the list of constraints.
 unify

